examples/features/authz/server/main.go

```diff
@@ -0,0 +1,265 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Binary server is an example server.
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/authz"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/examples/data"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+
+	pb "google.golang.org/grpc/examples/features/proto/echo"
+)
+
+const (
+	unaryEchoWriterRole      = "UNARY_ECHO:W"
+	streamEchoReaderRole     = "STREAM_ECHO:R"
+	streamEchoWriterRole     = "STREAM_ECHO:W"
+	streamEchoReadWriterRole = "STREAM_ECHO:RW"
+	authzPolicy              = `
+	{
+		"name": "authz",
+		"allow_rules": [
+			{
+				"name": "allow_UnaryEcho",
+				"request": {
+					"paths": ["/grpc.examples.echo.Echo/UnaryEcho"],
+					"headers": [
+						{
+							"key": "UNARY_ECHO:W",
+							"values": ["true"]
+						}
+					]
+				}
+			},
+			{
+				"name": "allow_ClientStreamingEcho",
+				"request": {
+					"paths": ["/grpc.examples.echo.Echo/ClientStreamingEcho"],
+					"headers": [
+						{
+							"key": "STREAM_ECHO:W",
+							"values": ["true"]
+						}
+					]
+				}
+			},
+			{
+				"name": "allow_ServerStreamingEcho",
+				"request": {
+					"paths": ["/grpc.examples.echo.Echo/ServerStreamingEcho"],
+					"headers": [
+						{
+							"key": "STREAM_ECHO:R",
+							"values": ["true"]
+						}
+					]
+				}
+			},
+			{
+				"name": "allow_BidirectionalStreamingEcho",
+				"request": {
+					"paths": ["/grpc.examples.echo.Echo/BidirectionalStreamingEcho"],
+					"headers": [
+						{
+							"key": "STREAM_ECHO:RW",
+							"values": ["true"]
+						}
+					]
+				}
+			}
+		],
+		"deny_rules": []
+	}
+	`
+)
+
+var (
+	port = flag.Int("port", 50051, "the port to serve on")
+
+	errMissingMetadata = status.Errorf(codes.InvalidArgument, "missing metadata")
+	errInvalidToken    = status.Errorf(codes.Unauthenticated, "invalid token")
+
+	mockedMetadata = newMockedMetadata()
+)
+
+func newMockedMetadata() metadata.MD {
+	md := metadata.MD{}
+	roles := []string{
+		unaryEchoWriterRole,
+		streamEchoReadWriterRole,
+	}
+	for _, role := range roles {
+		md.Set(role, "true")
+	}
+	return md
+}
```

Comment:
Instead of injecting these at the server interceptor, would it make more sense to inject appropriate roles at the client interceptor? That way, we will be able to illustrate both authorized and unauthorized RPCs.

Another option would be to include a "username" metadata in the RPC sent out by the client, and use that here and inject an appropriate role. That would also help with illustrating authorized and unauthorized RPCs.

Reply:
I quite like the second suggestion, I have injected a username in the client token, and made a `newContextWithRoles(ctx, username` implementation, which is later injected in the unary/stream auth interceptors